Show proper error to user when conf.py is not found

humitos opened the pull request:
I'm using BuildEnvironmentError for now, but we should use
ProjectConfigurationError once the PR got merged: https://github.com/rtfd/readthedocs.org/pull/3310

Closes #2963

agjohnson commented:
I'll wait until #3310 is merged and this is rebased on there for review. Blocking for now -- or feel free to rebase there and we'll swap the PR base after merge of #3310.

humitos commented:
I added a couple of tests for old behaviour and also to catch this PR behaviour. Besides, I added the exception message as a class attribute.

Please, re-review this PR when you have some time.

(the last commit is styling, so you can skip it from the logic review)

humitos commented:
Link for logic commits: https://github.com/rtfd/readthedocs.org/pull/3326/files/f44ba0eb3714c7e607925c75d1eec9c65fbe4a35

ericholscher commented on readthedocs/projects/exceptions.py:
```diff
@@ -15,6 +15,12 @@ class ProjectConfigurationError(BuildEnvironmentError):
+        "There are more than one conf.py file and none of them say doc "
+        "in their path, we don't know which one use. Please, select "
+        "the correct one under the Advanced settings tab in the "
+        "project's Admin."
```
We should probably provide a link here.

humitos replied:
Do we support HTML content on the message exception? I think no.

For example, we are not using `| safe` filter at https://github.com/rtfd/readthedocs.org/blob/bb7bb4d6751ecce67ba274a32100406b951f9b90/readthedocs/templates/builds/build_detail.html#L156

ericholscher replied:
Would make for a better UX, perhaps as an idea in the future. We could also add it as a "build tip"

agjohnson replied:
I like that idea. I think we need to firm up our usage around build tips and not do the logic in the template though.

ericholscher commented on readthedocs/projects/models.py:
```diff
@@ -529,8 +529,18 @@ def conf_file(self, version=LATEST):
+            raise ProjectConfigurationError(
+                ProjectConfigurationError.MULTIPLE_CONF_FILES
+            )
+
```
I feel like we keep changing build logic that breaks people's build. Is there a reason we are wanting to **fail** builds on this, instead of warning users? I think we probably want to warn them in this case, because changing this logic could easily break hundreds of project's builds for no real reason.

We did this with the `setup.py install` failures, and I don't think it's a good pattern to follow.

humitos replied:
I didn't change the logic here. I just added a better message.

Before this PR, it said "Conf.py not found" but there were more than one `conf.py` file. Now, it will fail as it was failing before but saying "Hey, too many conf.py files. We don't know which one to pick"

ericholscher replied:
hmm, interesting.

agjohnson commented on readthedocs/projects/exceptions.py:
```diff
@@ -8,13 +11,20 @@
+        'There are more than one conf.py file and none of them say doc '
+        "in their path, we don't know which one use. Please, select "
+        'the correct one under the Advanced settings tab in the '
+        "project's Admin."
```
A few notes on copy here:

- More clarity around *why* this is a problem is good -- that is: "We found more than one conf.py and are not sure which one to use." instead of mentioning `docs/`
- "Please, select the correct one" -> "Please specify the correct file"

humitos commented:
This PR should be ready to merge.

The linting error is unrelated.

```
File "/home/travis/build/rtfd/readthedocs.org/.tox/lint/lib/python2.7/site-packages/pylint_django/plugin.py", line 22, in register

    start = name_checker.config.const_rgx.pattern[:-2]

AttributeError: 'NoneType' object has no attribute 'pattern'
```

I didn't find _why_ it's failing yet.

stsewd commented on readthedocs/projects/models.py:
```diff
@@ -171,8 +172,8 @@ class Project(models.Model):
 
     # Other model data.
     path = models.CharField(_('Path'), max_length=255, editable=False,
```
In some parts the function call is in this way 
```python
func(arg1, 
        arg2)
```

and other parts like this

```python
func(
    arg1,
    arg2)
```

Which style is the _correct_?

humitos replied:
The second one, but with a trailing comma and the ´)´ in the folowing line is closer to what we want. Like this

```
func(
    arg1,
    arg2,
)
```

but we use this only when it doesn't fit in 80 columns.

We have a style guide for this: http://docs.readthedocs.io/en/latest/contribute.html?highlight=contributing#contributing-to-development

We are using `pre-commit` command that runs a couple of scripts to apply the proper style to the code.

humitos commented:
The linting error on this PR is solved in #3408

humitos commented:
I think this PR is read to merge. Can you review it again, please?
